### PR TITLE
nft-817 (Internal)

### PR DESCRIPTION
**Pull request by cawfree**
- Migrated the reNFT SDK to support the concept of versioned deployments.
- Contracts are now instantiated using a `RenftDeployment` object, which declares:
  - The network it was deployed on.
  - The address of the contract.
  - The type of the contract.
  - The version of the contract.
- Using this information, we are able to determine the correct type-safe interface to use for the contract.
  - This enabled us to change client-side contract interfaces arbitrarily, and elevate typing errors if they attempt to invoke the contract with incorrect data for a given version.

As part of this refactor, we needed to export an architecture which conforms to the convetions of how contracts with arbitrary interfaces may be deployed. This meant that:

- We moved away from using `class`es to define implementation logic. This is because `class`es expect a common interaction interface for subclasses, which is untrue for our intended use case.
- Functions which erroneously coupled the network and contract type, like `toScaledAmount` or `fromScaledAmount`, have been renamed to `toWhoopiScaledAmount` and `fromWhoopiScaledAmount`, since the original behavior was that for any network other than `whoopi`, we would `throw`. This allows us to still pass in a network to help define the appropriate `Resolver`, and also extends the possibility for `whoopi` to be deployed on other networks (even though we likely won't, it's just nice that the codebase reasonably permits this capability).
- To create a `Contract`, we call `getRenftContract` and pass an instance of a `RenftDeployment` to resolve the specific interface required for interaction.

### `dist/`

We are currently versioning the `dist/` directory on this branch to serve to the `front/`; before merging this branch we need to re-enable gitignoring the `dist/` directory prior to pushing to `npm`.

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -1,3 +1,19 @@
+export enum EthereumNetworkType {
```
Comment (nazariyv) [inline src/types.ts:1]: probably `EVMNetworkType` and then `EVMNetworkLike`?

File: src/types.ts [hunk 2]
```diff
@@ -18,6 +34,68 @@ export enum NFTStandard {
   E1155 = 1,
 }
 
+// A RenftContractType represents a unique "style" of rental functionality:
+//
+// + Azrael: A collateralized rental; an up-front fee is required to prevent a renter from leaving a lender empty-handed.
+// + Sylvester: A non-collateralized rental; have an escrow contract programmatically safeguard the safety of your asset.
+// + Whoopi A non-collateralized rental which is cycle-driven, and includes revenue-share for counterparties.
```
Comment (nazariyv) [inline src/types.ts:41]: `Whoopi: A non [...]`

Comment (nazariyv) [reply src/types.ts:41]: cycle-driven is a convention that does not touch the contract itself. Let's use this description:

`Whoopi: A collateral-free rental; very similar to Sylvester, with an addition of whitelisting and reward share features that facilitate guild-scholar management. Push approach to pay function.`

File: src/types.ts [hunk 3]
```diff
@@ -1,3 +1,19 @@
+export enum EthereumNetworkType {
+  ETHEREUM_MAINNET = 'ETHEREUM_MAINNET',
+  POLYGON_MAINNET = 'POLYGON_MAINNET',
+  AVALANCHE_MAINNET = 'AVALANCHE_MAINNET',
+  AVALANCHE_FUJI_TESTNET = 'AVALANCHE_FUJI_TESTNET',
+}
+
+export type EthereumNetworkLike<T extends EthereumNetworkType> = {
```
[the remark this answers is not in the document]
Comment (nazariyv) [reply src/types.ts:8]: not even sure `EVMNetworkLike` is good name for it. If it's not EVM, our contracts won't run. So it has to be `EVM`.

**Comment by cawfree** [comment]
`size` can be expected to fail since we're now versioning the `dist/` directory to support serving the branch to front since we have decided to forego deployment onto npm until we're satisfied with our internal usage.

**Comment by cawfree** [comment]
Some of our internal builds currently rely upon the presence of the `dist/` directory in this branch. 

Closing this branch in favour of #6, which will allow us to merge onto `main/` and release an up-to-date SDK without invalidating current internal clients. Once these have been refactored back to consuming from npm, we can delete this branch altogether.

**Pull request closed without merging**